src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs

```diff
@@ -830,10 +878,101 @@ internal static void EncryptKey(SafeKeyHandle safeKeyHandle, byte[] pbKey, int c
             Array.Reverse(pbEncryptedKey);
         }
 
+        internal static int EncryptData(
+            SafeKeyHandle hKey,
+            byte[] input,
+            int inputOffset,
+            int inputCount,
+            byte[] output,
+            int outputOffset,
+            int outputCount,
+            bool isFinal)
+        {
+            VerifyValidHandle(hKey);
+            Debug.Assert(input != null);
+            Debug.Assert(inputOffset >= 0);
+            Debug.Assert(inputCount >= 0);
+            Debug.Assert(inputCount <= input.Length - inputOffset);
+            Debug.Assert(output != null);
+            Debug.Assert(outputOffset >= 0);
+            Debug.Assert(outputCount >= 0);
+            Debug.Assert(outputCount <= output.Length - outputOffset);
+            Debug.Assert((inputCount % 8) == 0);
+            Debug.Assert((outputCount % 8) == 0);
+
+            // Figure out how big the encrypted data will be
+            // REVIEW NOTE: we could just bump up the buffer by 8 bytes always instead of doing the CryptEncrypt check for buffer size
+            int cbEncryptedData = inputCount;
+            if (!Interop.CryptEncrypt(hKey, SafeHashHandle.InvalidHandle, isFinal, 0, null, ref cbEncryptedData, cbEncryptedData))
+            {
+                throw new CryptographicException(SR.Format(SR.CryptEncrypt_Failed, Convert.ToString(GetErrorCode())));
+            }
+
+            // encryptedData is an in/out buffer for CryptEncrypt. Allocate space for the encrypted data, and copy the
+            // plaintext data into that space.  Since encrypted data will have padding applied, the size of the encrypted
+            // data should always be larger than the plaintext key, so use that to determine the buffer size.
+            Debug.Assert(cbEncryptedData >= inputCount);
+            var encryptedData = new byte[cbEncryptedData];
+            Buffer.BlockCopy(input, inputOffset, encryptedData, 0, inputCount);
+
+            // Encrypt for real - the last parameter is the total size of the in/out buffer, while the second to last
+            // parameter specifies the size of the plaintext to encrypt.
+            int encryptedDataLength = inputCount;
+            if (!Interop.CryptEncrypt(hKey, SafeHashHandle.InvalidHandle, isFinal, 0, encryptedData, ref encryptedDataLength, cbEncryptedData))
+            {
+                int errCode = GetErrorCode();
+                throw new CryptographicException(SR.CryptEncrypt_Failed, Convert.ToString(errCode));
+            }
+            Debug.Assert(encryptedDataLength == cbEncryptedData);
+
+            // If isFinal, padding was added so ignore it by using outputCount as size
+            Buffer.BlockCopy(encryptedData, 0, output, outputOffset, outputCount);
```

Comment:
isFinal does get set to true in this implementation; the netfx c++ also sets true and similarly removes the added padding